lib/_http_agent.js

```diff
@@ -112,11 +112,16 @@ function Agent(options) {
         if (this.sockets[name])
           count += this.sockets[name].length;
 
-        if (count > this.maxSockets || freeLen >= this.maxFreeSockets) {
-          socket.destroy();
-        } else if (this.keepSocketAlive(socket)) {
-          freeSockets = freeSockets || [];
-          this.freeSockets[name] = freeSockets;
+        if (this.keepSocketAlive(socket) &&
+            this.maxFreeSockets > 0 &&
+            count <= this.maxSockets) {
+          if (freeLen >= this.maxFreeSockets) {
+            const oldest = this.freeSockets[name].shift();
+            oldest.destroy();
+          } else {
```

Comment:
nit, else if (!freeSockets)

Reply:
I'm not quite following, could you explain a bit more.

Reply:
the block inside of the `else` is for the case when !freeSockets, could be simplified

Reply:
it's a nit, no biggie